In [1]:
import numpy as np
import pandas as pd
import sklearn
import matplotlib.pyplot as plt
import re
from geopy.geocoders import Nominatim, Photon
import math

### PLAN

1. Pobranie danych
2. Przejrzeć co mamy
    - wielkosc data setu
    - jakie kolumny
    - jakiego typu dane
    - podejrzec probke
    - jesli dane sa we wlasciwym formacie: proste statystyki
3. Konwersja danych na odpowiedni format
4. Co z brakujacymi wartosciami?

#### Zaimportuj i przejrzyj dane

In [2]:
dane = pd.read_csv('data/DataWorkshop.csv')

In [3]:
print('Rozmiar danych: ', dane.shape)

Rozmiar danych:  (229929, 30)


In [4]:
print('Kolumny w datasecie: ')
dane.columns

Kolumny w datasecie: 


Index(['Data zdarzenia', 'Godzina', 'GPS x', 'GPS y', 'Miejscowość', 'Ulica',
       'Numer domu', 'Numer drogi', 'KM HM (Pikietaż)', 'Obszar',
       'Skrzyżowanie', 'Geometria drogi', 'Charakterystyka miejsca zdarzenia',
       'Oświetlenie', 'Warunki atmosferyczne', 'Rodzaj zdarzenia', 'Wina',
       'Rodzaj drogi', 'Nawierzchnia drogi', 'Stan nawierzchni',
       'Sygnalizacja świetlna', 'Oznakowanie poziome',
       'Liczba pasów w zdarzeniu', 'Zachowanie kierującego',
       'Zachowanie pieszego', 'Inne przyczyny', 'Liczba rannych w zdarzeniu',
       'Liczba zabitych w zdarzeniu', 'Liczba kolizji', 'Liczba wypadków'],
      dtype='object')

In [5]:
print('Typy danych: ')
dane.dtypes

Typy danych: 


Data zdarzenia                        object
Godzina                               object
GPS x                                 object
GPS y                                 object
Miejscowość                           object
Ulica                                 object
Numer domu                            object
Numer drogi                           object
KM HM (Pikietaż)                      object
Obszar                                object
Skrzyżowanie                          object
Geometria drogi                       object
Charakterystyka miejsca zdarzenia     object
Oświetlenie                           object
Warunki atmosferyczne                 object
Rodzaj zdarzenia                      object
Wina                                  object
Rodzaj drogi                          object
Nawierzchnia drogi                    object
Stan nawierzchni                      object
Sygnalizacja świetlna                 object
Oznakowanie poziome                   object
Liczba pas

In [6]:
print('Brakujace wartosci: ')

missing = dane.isnull().sum()
print(missing)

Brakujace wartosci: 
Data zdarzenia                            0
Godzina                                   0
GPS x                                     0
GPS y                                     0
Miejscowość                               0
Ulica                                 20728
Numer domu                            93884
Numer drogi                               0
KM HM (Pikietaż)                     194888
Obszar                                    1
Skrzyżowanie                         166779
Geometria drogi                       66008
Charakterystyka miejsca zdarzenia         2
Oświetlenie                               0
Warunki atmosferyczne                     1
Rodzaj zdarzenia                          0
Wina                                      0
Rodzaj drogi                              1
Nawierzchnia drogi                        1
Stan nawierzchni                          1
Sygnalizacja świetlna                     0
Oznakowanie poziome                       2
Liczba pasó

In [7]:
print('Brakujace wartosci jako procent zbioru, %: ')
index=[]
lst=[]
for item in missing.iteritems():
    index.append(item[0])
    lst.append(round((item[1]/dane.shape[0])*100,4))
missing_percentage=pd.Series(lst, index=index)
print(missing_percentage)

Brakujace wartosci jako procent zbioru, %: 
Data zdarzenia                        0.0000
Godzina                               0.0000
GPS x                                 0.0000
GPS y                                 0.0000
Miejscowość                           0.0000
Ulica                                 9.0150
Numer domu                           40.8317
Numer drogi                           0.0000
KM HM (Pikietaż)                     84.7601
Obszar                                0.0004
Skrzyżowanie                         72.5350
Geometria drogi                      28.7080
Charakterystyka miejsca zdarzenia     0.0009
Oświetlenie                           0.0000
Warunki atmosferyczne                 0.0004
Rodzaj zdarzenia                      0.0000
Wina                                  0.0000
Rodzaj drogi                          0.0004
Nawierzchnia drogi                    0.0004
Stan nawierzchni                      0.0004
Sygnalizacja świetlna                 0.0000
Oznakowanie

In [8]:
# Probka danych
dane.sample(50)

Data zdarzenia   Godzina      GPS x      GPS y           Miejscowość  \
43514      2016-11-02  20:35:00  18*53'413  50*09'526               MIKOŁÓW   
38192      2016-10-03  11:20:00  19*09'461  50*34'375             RZENISZÓW   
35243      2016-09-15  15:10:00  18*38'380  50*19'081               GLIWICE   
207498     2019-08-29  07:39:00  18*57'264  50*17'216               CHORZÓW   
9425       2016-03-13  16:40:00  19*07'471  49*50'192                  KOZY   
133141     2018-05-28  12:30:00  19*08'079  50*14'249             MYSŁOWICE   
34832      2016-09-08  15:00:00  18*41'109  49*45'503                 GUMNA   
7793       2016-03-02  08:55:00  18*39'556  50*17'411               GLIWICE   
120543     2018-03-07  06:50:00  19*06'310  50*17'189             SOSNOWIEC   
136615     2018-06-20  18:00:00  19*07'514  50*48'421           CZĘSTOCHOWA   
149840     2018-09-14  15:45:00  19*01'264  50*16'029              KATOWICE   
209105     2019-09-06  19:20:00  18*52'136  50*26'101       TARNOWSKIE GÓRY   
124500     2018-04-03  09:00:00  19*10'158  49*40'296                ŻYWIEC   
187848     2019-04-28  15:10:00  19*00'299  50*16'054              KATOWICE   
132813     2018-05-26  21:48:00  19*02'215  50*53'126                 BIAŁA   
15721      2016-04-29  18:00:00  19*04'547  50*49'213           CZĘSTOCHOWA   
208105     2019-09-02  11:15:00  19*05'246  49*50'313         BIELSKO-BIAŁA   
4691       2016-02-04  17:40:00  18*47'298  49*48'114               SKOCZÓW   
190296     2019-05-13  08:10:00  19*02'222  49*49'145         BIELSKO-BIAŁA   
56523      2017-01-24  13:50:00  18*27'176  49*58'450      WODZISŁAW ŚLĄSKI   
103808     2017-11-22  08:50:00  19*07'123  50*21'225                BĘDZIN   
199981     2019-07-10  18:00:00  19*06'440  50*48'223           CZĘSTOCHOWA   
169007     2019-01-03  16:30:00  19*03'354  50*16'165              KATOWICE   
35256      2016-09-15  15:30:00  19*07'088  49*44'034            ŁODYGOWICE   
163782     2018-12-05  14:45:00  19*10'357  50*16'569             SOSNOWIEC   
93042      2017-09-22  10:15:00  18*55'570  50*17'218        ŚWIĘTOCHŁOWICE   
94671      2017-10-02  11:54:00  19*03'211  49*48'287         BIELSKO-BIAŁA   
76911      2017-06-10  14:25:00  18*36'309  50*13'194           NIEBOROWICE   
203294     2019-07-31  20:05:00  19*13'419  49*51'392               CZANIEC   
141436     2018-07-21  07:15:00  19*08'520  50*50'251           CZĘSTOCHOWA   
130187     2018-05-10  12:08:00  19*14'574  50*12'116              JAWORZNO   
213788     2019-10-03  14:20:00  18*38'054  49*45'002               CIESZYN   
86511      2017-08-15  22:30:00  19*00'150  49*54'170  CZECHOWICE-DZIEDZICE   
40981      2016-10-19  15:05:00  19*12'378  50*20'064      DĄBROWA GÓRNICZA   
9579       2016-03-15  08:30:00  18*37'159  50*24'038             PYSKOWICE   
114773     2018-01-26  09:58:00  19*08'576  50*17'129             SOSNOWIEC   
73379      2017-05-20  12:03:00  19*06'487  50*49'310           CZĘSTOCHOWA   
146291     2018-08-24  10:20:00  19*05'170  49*45'016             WILKOWICE   
196417     2019-06-14  14:30:00  19*07'529  50*16'305             SOSNOWIEC   
66799      2017-04-06  16:55:00  19*27'092  50*29'419             ZAWIERCIE   
115661     2018-02-02  09:35:00  18*58'587  50*14'070              KATOWICE   
85834      2017-08-10  12:30:00  19*03'006  49*48'099         BIELSKO-BIAŁA   
90406      2017-09-08  07:42:00  19*15'463  50*12'214              JAWORZNO   
58390      2017-02-06  11:15:00  19*13'034  50*20'125      DĄBROWA GÓRNICZA   
13535      2016-04-14  14:00:00  19*11'058  50*19'243      DĄBROWA GÓRNICZA   
103923     2017-11-22  16:40:00  18*58'447  50*12'006              KATOWICE   
186713     2019-04-20  20:00:00  18*43'528  50*18'341                ZABRZE   
197856     2019-06-27  08:00:00  18*13'238  50*05'319              RACIBÓRZ   
165759     2018-12-15  23:00:00  19*07'077  50*48'426           CZĘSTOCHOWA   
34983      2016-09-12  15:40:00  19*01'439  49*51'067    

### Konwersja typu danych

In [9]:
# Utworzenie kopii zbioru danych. Dalsza konwersja odbywac sie bedzie na kopii danych
df=dane.copy()
df.rename(columns={'Data zdarzenia': 'Data', 'GPS x': 'GPSx', 'GPS y': 'GPSy', 'Miejscowość':'Miejscowosc',
                   'Numer domu':'NrDomu','Numer drogi':'NrDrogi','KM HM (Pikietaż)':'Pikietaz', 'Skrzyżowanie':'Skrzyzowanie',
                  'Geometria drogi': 'GeometriaDrogi', 'Charakterystyka miejsca zdarzenia': 'CharMiejsca',
                  'Oświetlenie':'Oswietlenie', 'Warunki atmosferyczne':'Pogoda', 'Rodzaj zdarzenia':'RodzajZdarzenia',
                  'Rodzaj drogi':'RodzajDrogi', 'Nawierzchnia drogi':'NawierzchniaDrogi', 'Stan nawierzchni': 'StanNawierzchni',
                  'Sygnalizacja świetlna':'SygnalizacjaSwietlna', 'Oznakowanie poziome':'ZnakiPoziome',
                  'Liczba pasów w zdarzeniu':'LiczbaPasow','Zachowanie kierującego':'ZachowanieKierowcy',
                  'Zachowanie pieszego':'ZachowaniePieszego', 'Inne przyczyny':'InnePrzyczyny', 'Liczba rannych w zdarzeniu':'Ranni',
                  'Liczba zabitych w zdarzeniu':'Zabici', 'Liczba kolizji':'Kolizja', 'Liczba wypadków':'Wypadek'}, inplace=True)


#### Data i czas

In [10]:
# Data zdarzenia i Godzina scalone do jednej kolumny
print('kolumna DATA ZDARZENIA: ', df['Data'].dtypes)
print('kolumna GODZINA: ', df['Data'].dtypes)
df['data_full'] = df['Data'] + ' ' + df['Godzina']
df.sample(10)

kolumna DATA ZDARZENIA:  object
kolumna GODZINA:  object


Data   Godzina       GPSx       GPSy   Miejscowosc  \
96653   2017-10-11  15:40:00  19*07'081  50*18'120        BĘDZIN   
182625  2019-03-28  08:40:00  19*14'288  50*28'485      SIEWIERZ   
154994  2018-10-14  14:25:00  18*53'020  49*37'056         WISŁA   
194635  2019-06-06  16:10:00  18*24'552  50*03'252     RYDUŁTOWY   
45186   2016-11-15  12:50:00  18*35'315  50*06'578        RYBNIK   
138969  2018-07-06  10:25:00  18*32'438  50*06'001        RYBNIK   
57655   2017-02-01  14:00:00  18*42'160  50*17'279       GLIWICE   
26878   2016-07-12  12:45:00  18*41'322  49*57'439     PAWŁOWICE   
11554   2016-03-31  17:05:00  18*32'422  50*06'429        RYBNIK   
198037  2019-06-27  22:00:00  18*36'093  49*53'531  ZEBRZYDOWICE   

                       Ulica NrDomu NrDrogi Pikietaz                Obszar  \
96653                    NaN    NaN      86     17,2     Obszar zabudowany   
182625              KIELECKA    NaN      78      NaN     Obszar zabudowany   
154994                GŁĘBCE    32D    W941      NaN     Obszar zabudowany   
194635    ROMUALDA TRAUGUTTA    NaN  P5024S      NaN     Obszar zabudowany   
45186             MIKOŁOWSKA    NaN    W925      NaN  Obszar niezabudowany   
138969              GLIWICKA    NaN  P7014S      NaN     Obszar zabudowany   
57655             PRZEWOZOWA     32  GMINNA      NaN     Obszar zabudowany   
26878                 KRUCZA     10  P4163S      NaN     Obszar zabudowany   
11554               GLIWICKA     45      I2      NaN     Obszar zabudowany   
198037  JULIUSZA SŁOWACKIEGO    58B    W937      8,7     Obszar zabudowany   

        ... ZnakiPoziome LiczbaPasow  \
96653   ...         Jest         NaN   
182625  ...         Jest         2.0   
154994  ...       Nie ma         2.0   
194635  ...         Jest         2.0   
45186   ...         Jest         NaN   
138969  ...         Jest         2.0   
57655   ...       Nie ma         1.0   
26878   ...       Nie ma         NaN   
11554   ...         Jest         NaN   
198037  ...         Jest         2.0   

                                 ZachowanieKierowcy ZachowaniePieszego  \
96653   Niezachowanie bezp. odleg. między pojazdami                NaN   
182625        Nieustąpienie pierwszeństwa przejazdu                NaN   
154994                   Nieprawidłowe wyprzedzanie                NaN   
194635                      Nieprawidłowe wymijanie                NaN   
45186   Niezachowanie bezp. odleg. między pojazdami                NaN   
138969        Nieustąpienie pierwszeństwa przejazdu                NaN   
57655                         Nieprawidłowe cofanie                NaN   
26878                         Nieprawidłowe cofanie                NaN   
11554         Nieudzielenie pierwszeństwa przejazdu                NaN   
198037          Nieprawidłowe zmienianie pasa ruchu                NaN   

       InnePrzyczyny Ranni Zabici Kolizja Wypadek            data_full  
96653            NaN     0      0       1       0  2017-10-11 15:40:00  
182625           NaN     2      0       0       1  2019-03-28 08:40:00  
154994           NaN     1      0       0       1  2018-10-14 14:25:00  
194635           NaN     0      0       1       0  2019-06-06 16:10:00  
45186            NaN     0      0       1       0  2016-11-15 12:50:00  
138969           NaN     0      0       1       0  2018-07-06 10:25:00  
57655            NaN     0      0       1       0  2017-02-01 14:00:00  
26878            NaN     0      0       1       0  2016-07-12 12:45:00  
11554            NaN     0      0       1       0  2016-03-31 17:05:00  
198037           NaN     0      0       1       0  2019-06-27 22:00:00  

[10 rows x 31 columns]

In [11]:
# Data zdarzenia i Godzina do typu datetime
df['data_full'] = pd.to_datetime(df['data_full'], format='%Y-%m-%d %H:%M:%S')
df.sample(10)
# TO-DO: Jaki to dzien tygodnia? Jaki to dzien miesiaca? To dzien czy noc? Jaka faza ksiezyca itd..

Data   Godzina       GPSx       GPSy           Miejscowosc  \
169865  2019-01-05  13:25:00  19*08'529  50*22'138                SARNÓW   
29003   2016-08-02  05:45:00  18*51'240  50*08'595         ŁAZISKA GÓRNE   
116549  2018-02-04  19:45:00  19*11'290  50*15'422             SOSNOWIEC   
148160  2018-09-05  07:05:00  18*28'525  50*00'144      WODZISŁAW ŚLĄSKI   
161789  2018-11-23  18:50:00  18*41'320  50*02'491                  ŻORY   
64521   2017-03-22  08:10:00  18*42'458  50*16'026               GLIWICE   
31713   2016-08-20  17:00:00  19*07'278  50*13'322             MYSŁOWICE   
221980  2019-11-16  14:55:00  18*38'499  50*13'568                KNURÓW   
6100    2016-02-13  15:50:00  18*55'521  50*21'041                 BYTOM   
198207  2019-06-29  09:10:00  18*59'306  49*54'067  CZECHOWICE-DZIEDZICE   

                            Ulica NrDomu   NrDrogi Pikietaz  \
169865                     GŁÓWNA     23    P4719S      NaN   
29003                PSTROWSKIEGO     19    P5326S      NaN   
116549                        NaN    NaN        S1    542,3   
148160                  GRODZISKO      1    P5018S      NaN   
161789  MĘCZENNIKÓW OŚWIĘCIMSKICH     22    P9019S      NaN   
64521                  PSZCZYŃSKA    NaN        44      NaN   
31713          OBRZEŻNA ZACHODNIA    NaN    P6490S      NaN   
221980                  SZPITALNA     37  G370002S      NaN   
6100                    SZAFRANKA    NaN  G150218S      NaN   
198207                        NaN    NaN         1    593,9   

                      Obszar  ... ZnakiPoziome LiczbaPasow  \
169865     Obszar zabudowany  ...       Nie ma         2.0   
29003      Obszar zabudowany  ...         Jest         2.0   
116549  Obszar niezabudowany  ...         Jest         2.0   
148160  Obszar niezabudowany  ...         Jest         2.0   
161789     Obszar zabudowany  ...       Nie ma         2.0   
64521      Obszar zabudowany  ...         Jest         2.0   
31713   Obszar niezabudowany  ...         Jest         NaN   
221980     Obszar zabudowany  ...         Jest         3.0   
6100       Obszar zabudowany  ...         Jest         NaN   
198207  Obszar niezabudowany  ...         Jest         4.0   

                                 ZachowanieKierowcy ZachowaniePieszego  \
169865                        Nieprawidłowe cofanie                NaN   
29003         Nieustąpienie pierwszeństwa przejazdu                NaN   
116549  Niedostosowanie prędkości do warunków ruchu                NaN   
148160                      Nieprawidłowe wymijanie                NaN   
161789                          Gwałtowne hamowanie                NaN   
64521           Nieprawidłowe zmienianie pasa ruchu                NaN   
31713   Niezachowanie bezp. odleg. między pojazdami                NaN   
221980          Nieprawidłowe zmienianie pasa ruchu                NaN   
6100              Nieprzestrzeganie innych sygnałów                NaN   
198207  Niezachowanie bezp. odleg. między pojazdami                NaN   

       InnePrzyczyny Ranni Zabici Kolizja Wypadek           data_full  
169865           NaN     0      0       1       0 2019-01-05 13:25:00  
29003            NaN     0      0       1       0 2016-08-02 05:45:00  
116549           NaN     0      0       1       0 2018-02-04 19:45:00  
148160           NaN     0      0       1       0 2018-09-05 07:05:00  
161789           NaN     0      0       1       0 2018-11-23 18:50:00  
64521            NaN     0      0       1       0 2017-03-22 08:10:00  
31713            NaN     0      0       1       0 2016-08-20 17:00:00  
221980           NaN     0      0       1       0 2019-11-16 14:55:00  
6100             NaN     0      0       1       0 2016-02-13 15:50:00  
198207           NaN     0      0       1       0 2019-06-29 09:10:00  

[10 rows x 31 columns]

#### Wspolrzedne geograficzne

Domyslny zapis wspolrzednych to D\*M'S([D]egree, [M]inute, [S]cond) np. 18\*53'289. Jednak, gdy pierwsza cyfra sekund jest zero, czasem wystepuje zapis D\*M''S np. 18\*53''11. Wystepowaly takze bledne zapisy takie jak 4\*\*56'569 i 50\*09'\*01.

In [12]:
def dmsToDecimal(coordinate_as_string):
    splittedCoordinate=re.findall(r"\w+", coordinate_as_string)
    degrees = float(splittedCoordinate[0])
    minutes = float(splittedCoordinate[1])
    if len(splittedCoordinate)==3:
        seconds = float(splittedCoordinate[2])/10
    else:
        seconds = 0
    coordinate_as_float = degrees + minutes/60 + seconds/(60*60)
    return coordinate_as_float

In [13]:
# Przygotowana wyzej funkcje wywoluje na calych kolumnach dancych
df['GPSx'] = df['GPSx'].apply(lambda x: dmsToDecimal(x))
df['GPSy'] = df['GPSy'].apply(lambda x: dmsToDecimal(x))
df.sample(10)
# TO-DO Wyszukac i wykorzystac biblioteke okreslajaca wspolrzedne GPS na podstawie adresu
# TO-DO Popraw punkty ktore wypadly poza woj. slaskie bazujac na ulicy i numerze

Data   Godzina       GPSx       GPSy           Miejscowosc  \
113183  2018-01-17  08:20:00  18.962694  50.154583                 TYCHY   
51358   2016-12-21  10:10:00  18.640972  49.746028               CIESZYN   
130281  2018-05-11  06:45:00  18.916917  50.162972               MIKOŁÓW   
224671  2019-12-03  04:00:00  19.362583  50.923444              KŁOMNICE   
199024  2019-07-03  11:30:00  19.134861  50.281000             SOSNOWIEC   
130469  2018-05-11  16:35:00  18.953389  50.382500       PIEKARY ŚLĄSKIE   
134633  2018-06-07  09:50:00  19.005528  49.912917  CZECHOWICE-DZIEDZICE   
20804   2016-06-02  15:05:00  19.162889  50.768917           CZĘSTOCHOWA   
220601  2019-11-08  11:57:00  18.936722  50.264583               CHORZÓW   
163023  2018-12-01  18:25:00  18.909222  50.491694    MIASTECZKO ŚLĄSKIE   

                         Ulica NrDomu   NrDrogi Pikietaz  \
113183                PODLESKA     51    P8006S      NaN   
51358      BOLESŁAWA CHROBREGO    NaN    P2701S      NaN   
130281                 BIELSKA    NaN      W928      NaN   
224671           CZĘSTOCHOWSKA   110A        91      NaN   
199024                  3 MAJA    NaN    P7619S      NaN   
130469      GEN.JERZEGO ZIĘTKA     60  G280082S      NaN   
134633     JÓZEFA PIŁSUDSKIEGO      8    GMINNA      NaN   
20804   ALEJA WOJSKA POLSKIEGO    NaN         1    480,4   
220601        STEFANA BATOREGO    NaN    P9405S      NaN   
163023                METALOWA      8  G657004S      NaN   

                      Obszar  ... ZnakiPoziome LiczbaPasow  \
113183     Obszar zabudowany  ...       Nie ma         2.0   
51358      Obszar zabudowany  ...         Jest         NaN   
130281  Obszar niezabudowany  ...         Jest         4.0   
224671     Obszar zabudowany  ...         Jest         2.0   
199024     Obszar zabudowany  ...         Jest         2.0   
130469     Obszar zabudowany  ...         Jest         2.0   
134633     Obszar zabudowany  ...       Nie ma         2.0   
20804      Obszar zabudowany  ...         Jest         NaN   
220601     Obszar zabudowany  ...         Jest         2.0   
163023     Obszar zabudowany  ...       Nie ma         2.0   

                                 ZachowanieKierowcy ZachowaniePieszego  \
113183  Niedostosowanie prędkości do warunków ruchu                NaN   
51358         Nieustąpienie pierwszeństwa przejazdu                NaN   
130281                        Nieprawidłowe cofanie                NaN   
224671        Nieustąpienie pierwszeństwa przejazdu                NaN   
199024        Nieustąpienie pierwszeństwa przejazdu                NaN   
130469  Niezachowanie bezp. odleg. między pojazdami                NaN   
134633                        Nieprawidłowe cofanie                NaN   
20804     Niezachowanie bezp. odl. między pojazdami                NaN   
220601  Niezachowanie bezp. odleg. między pojazdami                NaN   
163023  Niedostosowanie prędkości do warunków ruchu                NaN   

       InnePrzyczyny Ranni Zabici Kolizja Wypadek           data_full  
113183           NaN     0      0       1       0 2018-01-17 08:20:00  
51358            NaN     0      0       1       0 2016-12-21 10:10:00  
130281           NaN     0      0       1       0 2018-05-11 06:45:00  
224671           NaN     0      0       1       0 2019-12-03 04:00:00  
199024           NaN     0      0       1       0 2019-07-03 11:30:00  
130469           NaN     0      0       1       0 2018-05-11 16:35:00  
134633           NaN     0      0       1       0 2018-06-07 09:50:00  
20804            NaN     0      0       1       0 2016-06-02 15:05:00  
220601           NaN     0      0       1       0 2019-11-08 11:57:00  
163023           NaN     0      0       1       0 2018-12-01 18:25:00  

[10 rows x 31 columns]

In [14]:
# Czy wszystkie punkty znajduja sie w granicach wojewodztwa slaskiego?
# Współrzędne geograficzne skrajnych punktów:
# północny: 51°05′57″ szer. geogr. N
# południowy: 49°23′38″ szer. geogr. N
# zachodni: 18°02′06″ dług. geogr. E
# wschodni: 19°58′26″ dług. geogr. E
northBorder = dmsToDecimal("51*05'57")
southBorder = dmsToDecimal("49*23'38")
westBorder = dmsToDecimal("18*02'06")
eastBorder = dmsToDecimal("19*58'26")

df[(df['GPSx']>eastBorder) | (df['GPSx']<westBorder) | (df['GPSy']>northBorder) | (df['GPSy']<southBorder)]

Data   Godzina       GPSx       GPSy       Miejscowosc  \
15349   2016-04-27  13:45:00  19.984250  50.269083          KATOWICE   
20566   2016-06-02  10:55:00  17.640861  50.335556           GLIWICE   
27105   2016-07-17  18:30:00  19.389639  51.554361             OSINY   
32758   2016-08-16  12:50:00  15.676500  50.771389         KONIECPOL   
32880   2016-08-30  16:00:00  21.999083  50.341028            ZABRZE   
34895   2016-09-09  14:00:00  16.775611  50.618917          BONOWICE   
37291   2016-09-28  19:45:00  15.998528  50.252694          KATOWICE   
45271   2016-11-16  07:20:00  19.995889  50.234028          KATOWICE   
49595   2016-12-10  23:12:00  19.345528  51.205611         SKRZYDLÓW   
50695   2016-12-16  19:05:00  19.315111  51.086361               ŁĘG   
52391   2016-12-25  21:40:00  18.930806  51.093472            LELITY   
58564   2017-02-07  14:42:00  18.930500  51.093583            LELITY   
86014   2016-11-18  15:50:00  19.340556  51.512694           MYKANÓW   
101039  2017-11-05  12:32:00  18.741333  51.096222           TRONINY   
120955  2018-03-09  08:10:00  19.250694  52.363667  DĄBROWA GÓRNICZA   
122869  2018-03-21  19:17:00  19.261250  55.012361          JAWORZNO   
131415  2018-05-17  09:15:00  19.166028  55.012000         MYSŁOWICE   
134588  2018-06-06  20:30:00  15.062306  50.269194          KATOWICE   
135253  2018-06-11  11:11:00  18.594722   4.949139  JASTRZĘBIE-ZDRÓJ   
139483  2018-07-09  21:35:00  18.529750  55.015167            RYBNIK   
145949  2018-08-22  09:45:00  19.985278  50.224806          KATOWICE   
146928  2018-08-28  13:30:00  19.982472  50.777972        BLACHOWNIA   
147956  2018-09-04  08:00:00  19.994528  50.238861          KATOWICE   
155437  2018-10-16  11:30:00  19.978639  50.277222          KATOWICE   
160399  2018-11-15  11:35:00  15.969667  50.223250          KATOWICE   
161478  2018-11-22  14:10:00  19.981639  50.187778          KATOWICE   
161915  2018-11-25  07:47:00  16.948417  50.272389           CHORZÓW   
162191  2018-11-26  19:20:00  19.047639  51.264417          KATOWICE   
173479  2018-11-17  07:40:00  15.971806  50.220139          KATOWICE   
203804  2019-08-04  12:03:00  19.825083  49.334194             WITÓW   

                                 Ulica NrDomu   NrDrogi Pikietaz  \
15349   TRASA NIKODEMA I JÓZEFA RENCÓW    NaN      W902      NaN   
20566                        LITERATÓW     69        I2      NaN   
27105                    CZĘSTOCHOWSKA     52       791      NaN   
32758                            RYNEK    NaN       794      NaN   
32880                              NaN    NaN        A4    318,1   
34895                              NaN    86A        78      NaN   
37291                              NaN    NaN        A4    335,9   
45271                        BRYNOWSKA     53    P6453S      NaN   
49595                           GŁÓWNA    NaN    GMINNA      NaN   
50695                              NaN    NaN         1    447,8   
52391                              NaN     16  G679047S      NaN   
58564                              NaN     15    P2030S      NaN   
86014                              NaN    NaN         1    462,3   
101039                             NaN     1A  G667053S      NaN   
120955               ŚW. JANA PAWŁA II    NaN         1    529,5   
122869                     GRUNWALDZKA    152    GMINNA      NaN   
131415                             NaN    NaN        A4    349,7   
134588        WALENTEGO ROŹDZIEŃSKIEGO    200        I2      NaN   
135253                    WIELKOPOLSKA     3A  G220071S      NaN   
139483                    JANA KOTUCZA    NaN        78      NaN   
145949                        KOLEJOWA     19        I2      NaN   
146928                   CZĘSTOCHOWSKA     59        46      NaN   
147956                       BRYNOWSKA     14    P6453S      NaN   
155437                     TYSIĄCLECIA     29    P6538S      NaN   
160399                     PIOTROWICKA    NaN    P6497S      NaN   
161478       

#### Miejsce zdarzenia



In [15]:
# Ulice zapisywane sa w sposob niespojny np. Mickiewicza, A.Mickiewicza, Adama Mickiewicza
# Czy sa zapisywane spojnie wzgledem miejscowosci?
#df[df['Ulica'].str.contains('MICKIEWICZA',na=False)&df['Miejscowosc'].str.contains('KATOWICE',na=False)]
df[df['Ulica'].str.contains('MICKIEWICZA',na=False)&df['Miejscowosc'].str.contains('CZĘSTOCHOWA',na=False)]
# Wniosek: Zapis nazwy ulicy nie jest spojny nawet w ramach jednej miejscowosci

Data   Godzina       GPSx       GPSy  Miejscowosc  \
3080    2016-01-22  12:40:00  19.117972  50.801083  CZĘSTOCHOWA   
4814    2016-02-07  10:05:00  19.113667  50.800972  CZĘSTOCHOWA   
4990    2016-02-08  18:30:00  19.115667  50.801167  CZĘSTOCHOWA   
5013    2016-02-02  16:30:00  19.070306  50.480306  CZĘSTOCHOWA   
5669    2016-02-12  17:30:00  19.118250  50.801333  CZĘSTOCHOWA   
8394    2016-03-03  16:15:00  19.117861  50.800694  CZĘSTOCHOWA   
14907   2016-04-21  14:55:00  19.113583  50.801000  CZĘSTOCHOWA   
18767   2016-05-17  12:35:00  19.117306  50.800667  CZĘSTOCHOWA   
19048   2016-05-20  09:35:00  19.117639  50.800778  CZĘSTOCHOWA   
20173   2016-05-27  12:10:00  19.117639  50.800806  CZĘSTOCHOWA   
21471   2016-06-07  10:30:00  19.118056  50.800944  CZĘSTOCHOWA   
24369   2016-06-26  10:40:00  19.115694  50.801139  CZĘSTOCHOWA   
32100   2016-08-19  11:00:00  19.117694  50.800639  CZĘSTOCHOWA   
36186   2016-09-17  23:35:00  19.116833  50.800861  CZĘSTOCHOWA   
49346   2016-12-09  13:50:00  19.117528  50.800833  CZĘSTOCHOWA   
53919   2017-01-05  14:38:00  19.119167  50.800111  CZĘSTOCHOWA   
54241   2017-01-09  07:44:00  19.117528  50.800861  CZĘSTOCHOWA   
56115   2017-01-20  18:30:00  19.114444  50.801028  CZĘSTOCHOWA   
74675   2017-05-28  20:25:00  19.117889  50.801139  CZĘSTOCHOWA   
78420   2017-06-21  08:32:00  19.115972  50.801194  CZĘSTOCHOWA   
82431   2017-07-17  13:35:00  19.117528  50.801056  CZĘSTOCHOWA   
93868   2017-09-27  15:34:00  19.113556  50.801000  CZĘSTOCHOWA   
110835  2017-12-31  13:54:00  19.113000  50.800917  CZĘSTOCHOWA   
111370  2018-01-04  14:53:00  19.117472  50.800861  CZĘSTOCHOWA   
151161  2018-09-22  11:15:00  19.117944  50.801167  CZĘSTOCHOWA   
152957  2018-10-03  08:10:00  19.114167  50.801028  CZĘSTOCHOWA   
162623  2018-11-29  14:50:00  19.114333  50.801056  CZĘSTOCHOWA   
176488  2019-02-15  15:35:00  19.116639  50.801139  CZĘSTOCHOWA   
184484  2019-04-08  13:10:00  19.113833  50.800944  CZĘSTOCHOWA   
188362  2019-05-01  17:00:00  19.115250  50.801056  CZĘSTOCHOWA   
189531  2019-05-09  14:40:00  19.116917  50.801056  CZĘSTOCHOWA   
207100  2019-08-26  14:30:00  19.114917  50.801056  CZĘSTOCHOWA   
218575  2019-10-28  17:50:00  19.114917  50.801056  CZĘSTOCHOWA   

                      Ulica NrDomu   NrDrogi Pikietaz             Obszar  ...  \
3080    RONDO A.MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
4814            MICKIEWICZA  25/31  G110525S      NaN  Obszar zabudowany  ...   
4990            MICKIEWICZA    NaN  G110525S      NaN  Obszar zabudowany  ...   
5013            MICKIEWICZA    NaN  G110525S      NaN  Obszar zabudowany  ...   
5669      RONDO MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
8394      RONDO MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
14907           MICKIEWICZA  25/31  G110525S      NaN  Obszar zabudowany  ...   
18767           MICKIEWICZA      1  G110525S      NaN  Obszar zabudowany  ...   
19048     RONDO MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
20173     RONDO MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
21471     RONDO MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
24369           MICKIEWICZA    NaN  G110525S      NaN  Obszar zabudowany  ...   
32100     RONDO MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
36186           MICKIEWICZA   5/17  G110525S      NaN  Obszar zabudowany  ...   
49346   RONDO A.MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
53919     RONDO MICKIEWICZA      1    P6614S      NaN  Obszar zabudowany  ...   
54241           MICKIEWICZA    NaN  G110525S      NaN  Obszar zabudowany  ...   
56115           MICKIEWICZA  25/31  G110525S      NaN  Obszar zabudowany  ...   
74675   RONDO A.MICKIEWICZA    NaN    P6614S      NaN  Obszar zabudowany  ...   
78420           MICKIEWICZA     12  G110525S      NaN  Obszar zabudowany  ...   
82431     RONDO MICKIEWICZA    NaN

In [16]:
# Pikietaz okresla miejsce zdarzenia na drodze poza terenem zabudowanym
#Zapisywany jako xx,yyy gdzie:
#xx – całkowita liczba kilometrów od początku drogi;
#yyy – całkowita liczba metrów liczona od ostatniego pełnego kilometra.
# W zbiorze danych prawdopodobnie jako wartosc w km, tyle ze z uzyciem "," jako separatora dziesietnego
def pikietazToFloat(pikietazRaw):
    if isinstance(pikietazRaw, str):
        pikietazAsFloat = float(pikietazRaw.replace(",", "."))
    else:
        pikietazAsFloat = pikietazRaw
    return pikietazAsFloat
df['Pikietaz']=df['Pikietaz'].apply(lambda x: pikietazToFloat(x))
df['Pikietaz']=df['Pikietaz'].fillna('')
df[df['Pikietaz'].notnull()].sample(10)

Data   Godzina       GPSx       GPSy       Miejscowosc  \
132519  2018-05-19  17:30:00  19.141472  50.300861         SOSNOWIEC   
73108   2017-05-18  17:55:00  18.649917  50.251722           GLIWICE   
39597   2016-10-11  01:40:00  19.049556  49.800806     BIELSKO-BIAŁA   
57873   2017-02-02  09:58:00  18.955694  50.295889           CHORZÓW   
175387  2019-02-08  23:50:00  19.096139  50.790111       CZĘSTOCHOWA   
88993   2017-09-01  08:15:00  18.854806  50.457583   TARNOWSKIE GÓRY   
126922  2018-04-18  17:11:00  19.217111  50.100028      CHEŁM ŚLĄSKI   
79687   2017-06-29  08:00:00  19.020222  50.254389          KATOWICE   
117331  2018-02-13  13:25:00  18.724028  49.961944        WARSZOWICE   
217272  2019-10-22  09:00:00  19.237528  50.318889  DĄBROWA GÓRNICZA   

                    Ulica NrDomu   NrDrogi Pikietaz                Obszar  \
132519                NaN     45        94              Obszar zabudowany   
73108            RYBNICKA    NaN        78           Obszar niezabudowany   
39597         PARTYZANTÓW    112      W942              Obszar zabudowany   
57873           KATOWICKA    NaN        79              Obszar zabudowany   
175387          KOŚCIELNA     86  G110391S              Obszar zabudowany   
88993             GRODZKA    NaN    P3291S              Obszar zabudowany   
126922        GÓRNOŚLĄSKA     40       780              Obszar zabudowany   
79687        JAGIELLOŃSKA      6    P6441S              Obszar zabudowany   
117331         CIESZYŃSKA      4        81           Obszar niezabudowany   
217272  ŚW. JANA PAWŁA II    NaN        S1      535  Obszar niezabudowany   

        ... ZnakiPoziome LiczbaPasow  \
132519  ...         Jest         2.0   
73108   ...         Jest         2.0   
39597   ...         Jest         NaN   
57873   ...         Jest         3.0   
175387  ...       Nie ma         2.0   
88993   ...         Jest         2.0   
126922  ...         Jest         2.0   
79687   ...         Jest         NaN   
117331  ...         Jest         2.0   
217272  ...         Jest         4.0   

                                 ZachowanieKierowcy ZachowaniePieszego  \
132519        Nieustąpienie pierwszeństwa przejazdu                NaN   
73108   Niezachowanie bezp. odleg. między pojazdami                NaN   
39597                                           NaN                NaN   
57873   Niezachowanie bezp. odleg. między pojazdami                NaN   
175387        Nieustąpienie pierwszeństwa przejazdu                NaN   
88993    Nieprzestrzeganie znaków i innych sygnałów                NaN   
126922                                          NaN                NaN   
79687           Nieprawidłowe zatrzymywanie, postój                NaN   
117331  Niezachowanie bezp. odleg. między pojazdami                NaN   
217272          Nieprawidłowe zmienianie pasa ruchu                NaN   

                       InnePrzyczyny Ranni Zabici Kolizja Wypadek  \
132519                           NaN     0      0       1       0   
73108                            NaN     0      0       1       0   
39597   Obiekty, zwierzęta na drodze     0      0       1       0   
57873                            NaN     0      0       1       0   
175387                           NaN     0      0       1       0   
88993                            NaN     0      0       1       0   
126922  Obiekty, zwierzęta na drodze     0      0       1       0   
79687                            NaN     1      0       0       1   
117331                           NaN     0      0       1       0   
217272                           NaN     0      0       1       0   

                 data_full  
132519 2018-05-19 17:30:00  
73108  2017-05-18 17:55:00  
39597  2016-10-11 01:40:00  
57873  2017-02-02 09:58:00  
175387 2019-02-08 23:50:00  
88993  2017-09-01 08:15:00  
126922 2018-04-18 17:11:00  
79687  2017-06-29 08:00:00  
117331 2018-02-13 13:25:00  
217272 2019-10-22 09:00:00  

[10 rows x 31 columns]

In [17]:
# Zamiana wartosci NaN na puste stringi. Ulatwi to korzystanie z bibliotek do geolokacji
df['Pikietaz']=df['Pikietaz'].fillna('')
df['Miejscowosc'] = df['Miejscowosc'].fillna('')
df['Ulica'] = df['Ulica'].fillna('')
df['NrDomu'] = df['NrDomu'].fillna('')
df['NrDrogi'] = df['NrDrogi'].fillna('')

In [18]:
# Obszar: zabudowany/niezabudowany
# Przypisane zostaly wartosci logiczne: Obszar zabudowany-True, Obszar niezabudowany False.
# W razie potrzeby takie wartosci mozna wygodnie przeksztalcic na int o wartosciach odpowiednio 1 i 0
df['Obszar'].value_counts()
df['Obszar'] = df['Obszar'].apply(lambda x: True if x=='Obszar zabudowany' else False)
df.sample(10)

Data   Godzina       GPSx       GPSy       Miejscowosc  \
63024   2017-03-07  19:50:00  18.321639  49.922417          CHAŁUPKI   
2880    2016-01-24  09:35:00  18.672556  49.910861    PIELGRZYMOWICE   
180862  2019-03-16  08:00:00  19.106250  50.817667       CZĘSTOCHOWA   
46826   2016-11-25  16:40:00  18.489833  50.046722            RADLIN   
96553   2017-10-10  19:30:00  18.551917  50.063306            RYBNIK   
26316   2016-07-12  12:15:00  18.594139  49.949083  JASTRZĘBIE-ZDRÓJ   
99215   2017-10-26  12:36:00  18.953528  49.989528          PSZCZYNA   
226860  2019-12-14  12:45:00  18.529194  50.085028            RYBNIK   
80814   2017-07-06  15:04:00  19.161361  50.598028        KOZIEGŁOWY   
129509  2018-05-06  16:30:00  18.973417  50.282306          KATOWICE   

                      Ulica NrDomu   NrDrogi Pikietaz  Obszar  ...  \
63024            BOGUMIŃSKA     26  G612903S             True  ...   
2880            GOLASOWICKA           P4162S             True  ...   
180862  ALEJA JANA PAWŁA II    106        I2             True  ...   
46826              MARIACKA         G530006S             True  ...   
96553                1 MAJA     55      W929             True  ...   
26316          WIELKOPOLSKA     2B        I2             True  ...   
99215                                      1             True  ...   
226860         WODZISŁAWSKA     90        78             True  ...   
80814             MONIUSZKI      1  G595070S             True  ...   
129509  BOLESŁAWA CHROBREGO     37    P6418S             True  ...   

       ZnakiPoziome LiczbaPasow                           ZachowanieKierowcy  \
63024          Jest         NaN                   Nieprawidłowe wyprzedzanie   
2880           Jest         NaN  Niedostosowanie prędkości do warunków ruchu   
180862       Nie ma         2.0                        Nieprawidłowe cofanie   
46826          Jest         2.0        Nieustąpienie pierwszeństwa przejazdu   
96553          Jest         NaN                               Inne przyczyny   
26316        Nie ma         NaN                       Nieprawidłowe omijanie   
99215          Jest         NaN          Nieprawidłowe zmienianie pasa ruchu   
226860         Jest         2.0  Niezachowanie bezp. odleg. między pojazdami   
80814        Nie ma         NaN                      Nieprawidłowe skręcanie   
129509         Jest         2.0          Nieprawidłowe zatrzymywanie, postój   

       ZachowaniePieszego InnePrzyczyny Ranni Zabici Kolizja Wypadek  \
63024                 NaN           NaN     0      0       1       0   
2880                  NaN           NaN     0      0       1       0   
180862                NaN           NaN     0      0       1       0   
46826                 NaN           NaN     0      0       1       0   
96553      Inne przyczyny           NaN     1      0       0       1   
26316                 NaN           NaN     0      0       1       0   
99215                 NaN           NaN     0      0       1       0   
226860                NaN           NaN     0      0       1       0   
80814                 NaN           NaN     0      0       1       0   
129509                NaN           NaN     0      0       1       0   

                 data_full  
63024  2017-03-07 19:50:00  
2880   2016-01-24 09:35:00  
180862 2019-03-16 08:00:00  
46826  2016-11-25 16:40:00  
96553  2017-10-10 19:30:00  
26316  2016-07-12 12:15:00  
99215  2017-10-26 12:36:00  
226860 2019-12-14 12:45:00  
80814  2017-07-06 15:04:00  
129509 2018-05-06 16:30:00  

[10 rows x 31 columns]

In [19]:
# Skrzyzowanie
# Zmiana nazwy na niezawierajaca polskich znakow
df['Skrzyzowanie'].unique()
df['Skrzyzowanie'].value_counts()
# Zamiana wartosci tekstowych na int zgodnie z zasada:
# 0 : O ruchu okrężnym; 1 : Równorzędne; 2 : Z drogą z pierwsz.
skrzyzowaniemap = {'O ruchu okrężnym':0, 'Równorzędne':1, 'Z drogą z pierwsz.':2}
df=df.applymap(lambda x: skrzyzowaniemap.get(x) if x in skrzyzowaniemap else x)

In [20]:
df['GeometriaDrogi'].value_counts()
# Geometria pozioma: 
# 0: Odcinek prosty; 1: Zakret(luk)

# Geometria pionowa: 
# 0: Spadek; 1: wzniesienie; 2: wierzcholek wzniesienia
# Skelic skrzyzowanie z geometria

Odcinek prosty                      142703
Zakręt, łuk                          16917
Spadek                                1866
Wzniesienie                           1317
Odcinek prosty; Spadek                 430
Odcinek prosty; Wzniesienie            329
Zakręt, łuk; Spadek                    179
Zakręt, łuk; Wzniesienie               108
Wierzchołek wzn.                        50
Odcinek prosty; Wierzchołek wzn.        16
Zakręt, łuk; Wierzchołek wzn.            5
Odcinek prosty; Zakręt, łuk              1
Name: GeometriaDrogi, dtype: int64

In [21]:
df['Geometria_drogi_poziom'] = df['GeometriaDrogi'].copy()
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 0 if 'Odcinek prosty' in str(x) else x)
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 1 if 'Zakręt' in str(x) else x)
df['Geometria_drogi_poziom'] = df['Geometria_drogi_poziom'].apply(lambda x: 1 if 'łuk' in str(x) else x)
# Co z reszta? Tzn z sytuacja kiedy nie jest zdefiniowane czy droga byla prosta czy byl to zakret?
df['Geometria_drogi_poziom'].value_counts()

0                   143479
1                    17209
Spadek                1866
Wzniesienie           1317
Wierzchołek wzn.        50
Name: Geometria_drogi_poziom, dtype: int64

In [22]:
df['Geometria_drogi_pion'] = df['GeometriaDrogi'].copy()
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 0 if 'Spadek' in str(x) else x)
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 1 if 'Wzniesienie' in str(x) else x)
df['Geometria_drogi_pion'] = df['Geometria_drogi_pion'].apply(lambda x: 2 if 'Wierzchołek wzn.' in str(x) else x)
# Co z reszta? Tzn z sytuacja kiedy nie jest zdefiniowane uksztaltowanie terenu?
df['Geometria_drogi_pion'].value_counts()

Odcinek prosty                 142703
Zakręt, łuk                     16917
0                                2475
1                                1754
2                                  71
Odcinek prosty; Zakręt, łuk         1
Name: Geometria_drogi_pion, dtype: int64

In [23]:
# Charakterystyka miejsca zdarzenia
df['CharMiejsca'].value_counts()

Jezdnia                                             191185
Parking, plac, MOP                                   17988
Pobocze                                               4676
Przejście dla pieszych                                3624
Chodnik, droga dla pieszych                           2612
Most, wiadukt, łącznica, tunel                        1796
Skarpa, rów                                           1356
Wjazd, wyjazd z posesji, pola                         1322
Roboty drogowe, oznakowanie tymczasowe                1235
Parking, plac, MOP (od 11.2015)                       1091
Przejazd kolejowy strzeżony                            554
Przejazd tramwajowy, torowisko tramwajowe              493
Parking, plac                                          479
Przystanek komunikacji publicznej                      376
Pas dzielący jezdnie                                   331
Przejazd dla rowerzystów                               289
Droga, pas ruchu, śluza dla rowerów                    2

In [24]:
# Unifikacja wartosci
df['CharMiejsca'] = df['CharMiejsca'].apply(lambda x: 'Parking' if 'Parking' in str(x) else x)
df['CharMiejsca'] = df['CharMiejsca'].apply(lambda x: 'Przejazd rowerowy' if 'Przejazd dla rowerzystów' in str(x) else x)
df['CharMiejsca'] = df['CharMiejsca'].apply(lambda x: 'Sluza dla rowerow' if 'Droga, pas ruchu, śluza dla rowerów' in str(x) else x)
df['CharMiejsca'].value_counts()

Jezdnia                                      191185
Parking                                       19558
Pobocze                                        4676
Przejście dla pieszych                         3624
Chodnik, droga dla pieszych                    2612
Most, wiadukt, łącznica, tunel                 1796
Skarpa, rów                                    1356
Wjazd, wyjazd z posesji, pola                  1322
Roboty drogowe, oznakowanie tymczasowe         1235
Przejazd kolejowy strzeżony                     554
Przejazd tramwajowy, torowisko tramwajowe       493
Przystanek komunikacji publicznej               376
Pas dzielący jezdnie                            331
Przejazd rowerowy                               302
Sluza dla rowerow                               272
Przejazd kolejowy niestrzeżony                  100
Przejazd tramwajowy, torowisko                   66
Przewiązka na drodze dwujezdniowej               50
Droga dla rowerzystów                            19
Name: CharMi

In [25]:
# Oświetlenie miejsca zdarzenia
df['Oswietlenie'].value_counts()
# Dane wystepuja bez powtorzen. Dodam kolumne dzien
df['CzyDzien'] = df['Oswietlenie'].apply(lambda x: True if 'Światło dzienne' in str(x) else False)
df['CzyDzien'].value_counts()

True     171386
False     58543
Name: CzyDzien, dtype: int64

In [26]:
# Warunki atmosferyczne
df['Pogoda'].value_counts()

Dobre warunki atmosferyczne                                   145165
Pochmurno                                                      43516
Opady deszczu                                                  23747
Opady śniegu, gradu                                             6338
Pochmurno; Opady deszczu                                        5104
Oślepiające słońce                                              2848
Mgła, dym                                                       1033
Pochmurno; Opady śniegu, gradu                                   741
Silny wiatr                                                      534
Silny wiatr; Opady deszczu                                       198
Opady deszczu; Opady śniegu, gradu                               158
Silny wiatr; Pochmurno                                           158
Pochmurno; Mgła, dym                                             121
Pochmurno; Opady deszczu; Opady śniegu, gradu                     73
Silny wiatr; Pochmurno; Opady desz

In [27]:
# Podzial na grupy:
# Pochmurno: True/False
df['Zachmurzenie'] = df['Pogoda'].apply(lambda x: True if 'Pochmurno' in str(x) else False)
# Deszcz: True/False
df['Deszcz'] = df['Pogoda'].apply(lambda x: True if 'Opady deszczu' in str(x) else False)
# Snieg/Grad: True/False
df['SniegGrad'] = df['Pogoda'].apply(lambda x: True if 'Opady śniegu, gradu' in str(x) else False)
# Oslepiajace slonce: True/False
df['OslepiajaceSlonce'] = df['Pogoda'].apply(lambda x: True if 'Oślepiające słońce' in str(x) else False)
# Silny wiatr: True/false
df['SilnyWiatr'] = df['Pogoda'].apply(lambda x: True if 'Silny wiatr' in str(x) else False)
# Mgla/dym: True/false
df['MglaDym'] = df['Pogoda'].apply(lambda x: True if 'Mgła, dym' in str(x) else False)

In [28]:
# Rodzaj zdarzenia-wyglada ok
df['RodzajZdarzenia'].value_counts()

Zderzenie pojazdów boczne              77669
Zderzenie pojazdów tylne               59469
Najechanie na pojazd unieruchomiony    29030
Inne                                   16655
Najechanie na zwierzę                  10453
Najechanie na pieszego                  6749
Najechanie na barierę ochronną          6107
Najechanie na słup, znak                5998
Zderzenie pojazdów czołowe              5714
Najechanie na dziurę, wybój, garb       5256
Wywrócenie się pojazdu                  3995
Najechanie na drzewo                    1974
Najechanie na zapore kolejową            580
Zdarzenie z pasażerem                    280
Name: RodzajZdarzenia, dtype: int64

In [29]:
# Wina - ok
df['Wina'].value_counts()

z winy kierującego             198463
z innych przyczyn               25503
współwina uczestników ruchu      3611
z winy pieszego                  2352
Name: Wina, dtype: int64

In [30]:
# Rodzaj drogi
df['RodzajDrogi'].value_counts()

Jednojezdniowa dwukierunkowa    160724
Dwie jezdnie jednokierunkowe     37692
Jednokierunkowa                  20812
Autostrada                        6751
Ekspresowa                        3949
Name: RodzajDrogi, dtype: int64

In [31]:
# Nawierzchnia drogi
df['NawierzchniaDrogi'].value_counts()

Twarda      229380
Gruntowa       548
Name: NawierzchniaDrogi, dtype: int64

In [32]:
# Stan nawierzchni
df['StanNawierzchni'].value_counts()
# Mokra
df['NawierzchniaMokra'] = df['StanNawierzchni'].apply(lambda x: True if 'Mokra' in str(x) else False)
# Sucha
df['NawierzchniaSucha'] = df['StanNawierzchni'].apply(lambda x: True if 'Sucha' in str(x) else False)
# Oblodzona
df['NawierzchniaOblodzona'] = df['StanNawierzchni'].apply(lambda x: True if ('Oblodzenie, zaśnieżenie' in str(x) or 'Oblodzona, zaśnieżona' in str(x)) else False)
# Kaluze, rozlewiska
df['NawierzchniaKaluze'] = df['StanNawierzchni'].apply(lambda x: True if 'Kałuże' in str(x) else False)
# Dziury, wyboje
df['NawierzchniaDziury'] = df['StanNawierzchni'].apply(lambda x: True if 'Dziury' in str(x) else False)
# Zanieczyszczona
df['NawierzchniaZanieczyszczona'] = df['StanNawierzchni'].apply(lambda x: True if 'Zanieczyszczona' in str(x) else False)
# Koleiny, garby
df['NawierzchniaKoleiny'] = df['StanNawierzchni'].apply(lambda x: True if 'Koleiny' in str(x) else False)

In [33]:
df['NawierzchniaKoleiny'].value_counts()

False    229667
True        262
Name: NawierzchniaKoleiny, dtype: int64

In [34]:
df['SygnalizacjaSwietlna'].value_counts()

Brak                207626
Jest, działa         21062
Jest, nie działa      1241
Name: SygnalizacjaSwietlna, dtype: int64

In [35]:
df['ZnakiPoziome'] = df['ZnakiPoziome'].apply(lambda x: True if 'Jest' in str(x) else False)
df['ZnakiPoziome'].value_counts()

True     178816
False     51113
Name: ZnakiPoziome, dtype: int64

In [36]:
df['LiczbaPasow'].value_counts()

2.0      101699
4.0       14937
3.0        9186
1.0        8090
6.0        3577
5.0        2537
8.0         671
7.0         433
10.0        166
12.0        131
9.0         117
50.0         85
11.0         25
20.0         20
18.0         16
13.0         16
14.0         13
16.0         11
70.0          9
0.0           8
60.0          5
30.0          4
40.0          4
90.0          3
66.0          2
22.0          2
17.0          2
250.0         1
67.0          1
24.0          1
15.0          1
25.0          1
32.0          1
Name: LiczbaPasow, dtype: int64

In [37]:
len(df['ZachowanieKierowcy'].unique())

162

In [38]:
df['ZachowanieKierowcy'].value_counts()

Niezachowanie bezp. odleg. między pojazdami                                                                42698
Nieustąpienie pierwszeństwa przejazdu                                                                      34350
Niedostosowanie prędkości do warunków ruchu                                                                23158
Nieprawidłowe cofanie                                                                                      21176
Nieprawidłowe omijanie                                                                                     13811
                                                                                                           ...  
Jazda bez wymaganego oświetlenia; Nieprawidłowe: wyprzedzanie                                                  1
Jazda bez wymaganych świateł; Niedostosowanie prędkości do warunków ruchu                                      1
Gwałtowne hamowanie; Nieustąpienie pierwszeństwa przejazdu                                      

In [39]:
def findRepeatingSubstring(string):
    if string != 'nan':
        for x in range(1, len(string)):
            substring = string[:x]

            if substring * (len(string)//len(substring))+(substring[:len(string)%len(substring)]) == string:
                return re.sub(r"\W+", ' ', substring).rstrip()

In [40]:
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: findRepeatingSubstring(str(x)).lower().replace(":", "") if findRepeatingSubstring(str(x)) is not None else str(x).lower().replace(":",""))


df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'Inne przyczyny' if str(x)=='Inne przyczyny (od 11.2015)' else x)
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'Niestosowanie się do sygnalizacji świetlnej' if str(x)=='Niestosowanie się do sygnalizacji świetlnej (od 11.2015)' else x)
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'Nieustąpienie pierwszeństwa pieszemu na przejściu dla pieszych' if str(x)=='Nieustąpienie pierwszeństwa pieszemu na przejściu dla pieszych (od 11.2015)' else x)
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'Nieustąpienie pierwszeństwa pieszemu w innych okolicznościach' if str(x)=='Nieustąpienie pierwszeństwa pieszemu w innych okolicznościach (od 11.2015)' else x)
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'Nieustąpienie pierwszeństwa pieszemu przy skręcaniu w drogę poprzeczną' if str(x)=='Nieustąpienie pierwszeństwa pieszemu przy skręcaniu w drogę poprzeczną  (od 11.2015)' else x)
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'Omijanie pojazdu przed przejściem dla pieszych' if str(x)=='Omijanie pojazdu przed przejściem dla pieszych (od 11.2015)' else x)
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'Wyprzedzanie pojazdu przed przejściem dla pieszych' if str(x)=='Wyprzedzanie pojazdu przed przejściem dla pieszych  (od 11.2015)' else x)

df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: str(x).replace(" (od 11.2015)", ""))
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: str(x).replace(" od 11 2015", ""))

df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'wyprzedzanie pojazdu przed przejściem dla pieszych' if str(x)=='wyprzedzanie pojazdu przed przejściem dla pieszych ' else x)
df['ZachowanieKierowcy'] = df['ZachowanieKierowcy'].apply(lambda x: 'nieustąpienie pierwszeństwa pieszemu przy skręcaniu w drogę poprzeczną' if str(x)=='nieustąpienie pierwszeństwa pieszemu przy skręcaniu w drogę poprzeczną ' else x)

df['ZachowanieKierowcy'].value_counts().head(20)

niezachowanie bezp. odleg. między pojazdami                       42698
nieustąpienie pierwszeństwa przejazdu                             34414
nan                                                               27901
nieprawidłowe cofanie                                             23745
niedostosowanie prędkości do warunków ruchu                       23195
nieprawidłowe omijanie                                            14969
nieprawidłowe zmienianie pasa ruchu                               13870
nieprawidłowe skręcanie                                           10290
inne przyczyny                                                     9909
nieprawidłowe wymijanie                                            5825
niezachowanie bezp. odl. między pojazdami                          4540
nieprawidłowe wyprzedzanie                                         3782
nieudzielenie pierwszeństwa przejazdu                              3692
nieustąpienie pierwszeństwa pieszemu na przejściu dla pieszych  

In [41]:
len(df['ZachowanieKierowcy'].unique())

113

In [42]:
# Zachowanie pieszego
df['ZachowaniePieszego'] = df['ZachowaniePieszego'].apply(lambda x: 'Inne przyczyny' if 'Inne przyczyny' in str(x) else x)
df['ZachowaniePieszego'] = df['ZachowaniePieszego'].apply(lambda x: 'Nieostrożne wejście na jezdnię: przed jadącym pojazdem' if 'Nieostrożne wejście na jezdnię: przed jadącym pojazdem' in str(x) else x)
df['ZachowaniePieszego'] = df['ZachowaniePieszego'].apply(lambda x: 'Leżenie, siedzenie, klęczenie, stanie na jezdni' if ('Leżenie, siedzenie, klęczenie, stanie na jezdni' in str(x) or 'Stanie na jezdni, leżenie' in str(x)) else x)
df['ZachowaniePieszego'] = df['ZachowaniePieszego'].apply(lambda x: 'Nieostrożne wejście na jezdnię: zza pojazdu, przeszkody' if 'Nieostrożne wejście na jezdnię: zza pojazdu, przeszkody' in str(x) else x)
df['ZachowaniePieszego'] = df['ZachowaniePieszego'].apply(lambda x: 'Przekraczanie jezdni w miejscu niedozwolonym' if 'Przekraczanie jezdni w miejscu niedozwolonym' in str(x) else x)
df['ZachowaniePieszego'] = df['ZachowaniePieszego'].apply(lambda x: 'Chodzenie nieprawidłową stroną drogi' if 'Chodzenie nieprawidłową stroną drogi' in str(x) else x)
df['ZachowaniePieszego'] = df['ZachowaniePieszego'].apply(lambda x: 'Wejście na jezdnię przy czerwonym świetle' if 'Wejście na jezdnię przy czerwonym świetle' in str(x) else x)
df['ZachowaniePieszego'].value_counts()  

Nieostrożne wejście na jezdnię: przed jadącym pojazdem     1270
Nieostrożne wejście na jezdnię: zza pojazdu, przeszkody     362
Inne przyczyny                                              293
Przekraczanie jezdni w miejscu niedozwolonym                287
Wejście na jezdnię przy czerwonym świetle                   156
Chodzenie nieprawidłową stroną drogi                         60
Leżenie, siedzenie, klęczenie, stanie na jezdni              45
Zatrzymanie, cofnięcie się                                   12
Name: ZachowaniePieszego, dtype: int64

In [43]:
# Inne przyczyny
df['InnePrzyczyny'] = df['InnePrzyczyny'].apply(lambda x: 'Niesprawność techniczna pojazdu' if 'Niesprawność techniczna pojazdu' in str(x) else x)
df['InnePrzyczyny'] = df['InnePrzyczyny'].apply(lambda x: 'Utrata przytomności, śmierć kierującego' if 'Utrata przytomności, śmierć kierującego' in str(x) else x)
df['InnePrzyczyny'].value_counts()

Obiekty, zwierzęta na drodze                11603
Nieustalone                                  5876
Stan jezdni                                  4828
Inne                                         2010
Niewłaściwy stan jezdni                       512
Niesprawność techniczna pojazdu               341
Zabezpieczenie robót na drodze                316
Utrata przytomności, śmierć kierującego       266
Z winy pasażera                               218
Pożar pojazdu                                 117
Oślepienie przez inny pojazd lub słońce        71
Organizacja ruchu                              19
Nieprawidłowo zabezp. roboty drogowe           18
Nieprawidłowo działająca zapora kolej.          5
Nieprawidłowo działająca zapora, rogatka        2
Działanie sygnalizacji świetlnej                1
Name: InnePrzyczyny, dtype: int64

In [44]:
df['Ranni'].value_counts()

0     217211
1      10459
2       1581
3        424
4        162
5         50
6         22
7          7
9          5
8          4
26         1
12         1
11         1
10         1
Name: Ranni, dtype: int64

In [45]:
df['Zabici'].value_counts()

0    229059
1       822
2        43
3         4
4         1
Name: Zabici, dtype: int64

In [46]:
df['Kolizja'].value_counts()

1    216531
0     13398
Name: Kolizja, dtype: int64

In [47]:
df['Wypadek'].value_counts()

0    216531
1     13398
Name: Wypadek, dtype: int64

###Spotkanie 3: Zdarzenia w terenie (nie/)zabudowanym: gdzie najwiecej zdarzen, wypadkow, kolizji
Gestosc zaludnienia, powody wypadkow a dzien i godzina [odleglosc od centrum]
